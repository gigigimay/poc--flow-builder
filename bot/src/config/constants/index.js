export const BulletType = {
  DOT: 'âˆ™',
  DASH: '-',
}
