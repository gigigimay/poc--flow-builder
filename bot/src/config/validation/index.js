import i18n from 'services/i18n'

export const isAnswer = (key) => (text, { language, tenant } = {}) => {
  return i18n.getFixedT(language)(`${tenant}:${key}`) === text
}

export const isAnswerSplitted = (key) => (text, { language, tenant } = {}) => {
  return (
    i18n.getFixedT(language)(`${tenant}:${key}`) ===
    (text?.split ? text.split(' ')[0] : text)
  )
}
