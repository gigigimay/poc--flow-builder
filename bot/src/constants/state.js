export const ERROR = 'ERROR'
export const ANSWER = 'ANSWER'
export const RESUME = 'RESUME'
export const INVALID = 'INVALID'
export const NEXT = 'NEXT'
export const SKIP = 'SKIP'
export const ASK = 'ASK'
export const COMMAND = 'COMMAND'
export const BACK = 'BACK'
export const PAUSED = 'PAUSED'
export const RESUMED = 'RESUMED'
export const BROKEN = 'BROKEN'
export const ASSIGNED = 'ASSIGNED'
export const FALLBACK = 'FALLBACK'