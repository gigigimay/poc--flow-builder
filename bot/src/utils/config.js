import * as fp from '@appman/mac_modules/fp'

export const mappingTranslations = (clientFeatures = {}) => {
  const translationResources = {}

  Object.keys(clientFeatures).forEach((tenant) => {
    const featureValue = clientFeatures[tenant]
    const { translations } = featureValue
    translationResources[tenant] = translations
  })

  return translationResources
}

export const filterAOCClient = (clientsConfig = []) => {
  return clientsConfig.filter(
    (clientConfig) =>
      fp.has('aoc_config', clientConfig) && fp.has('aoc_features', clientConfig)
  )
}
