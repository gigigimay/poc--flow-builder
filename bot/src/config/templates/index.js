import * as fp from '@appman/mac_modules/fp'

import { LINE } from 'constants/channel'
import { TEXT } from 'constants/template'

const templates = {
    [TEXT]: require('./text'),
}

export const getTemplate = (type) => (props) => () => {
  const template = fp.get(type, templates)

  if (!template) {
    throw new Error(`Can't find ${type} template`)
  }

  if (fp.isFunction(template)) {
    return template(props)
  }

  return template
}

export const commandMessage = (command) => ({ session }) => {
  // TODO: refactor messaging so that command message should be same message type
  const messageType = session.channel === LINE ? 'flex' : 'template'
  return {
    type: messageType,
    command,
  }
}

export const text = getTemplate(TEXT)
