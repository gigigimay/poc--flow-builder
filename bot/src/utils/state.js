import * as fp from '@appman/mac_modules/fp'

import * as commands from 'config/constants/commands'

import i18n from 'services/i18n'
import { isAnswer } from 'config/validation'
import { appSyncClient } from 'services/app-sync'
import { updateCaseMutation } from 'schema/case/mutation'
import { getFeature } from 'config/features'
import { getLogger } from './logger'

const logger = getLogger('utils/state.js')

export const isTextMessage = fp.pipe(fp.get('message.type'), fp.isEqual('text'))
export const isDevelopmentMode = process.env.DEVELOPMENT_MODE === 'true'

export const getNextCommand = ({ actions }) => ({
  cond: isDevelopmentMode ? isAnswer('next') : () => false,
  actions,
})

/** only these keys will be translated when sending messages */
const translationKeys = [
  'text',
  'altText',
  'label',
  'title',
  'subtitle',
  'bodyMessage',
  'buttonTitle',
  'buttonMessage',
]

export const isYes = fp.isEqual('yes')

export const isNo = fp.isEqual('no')

export const getModule = fp.pipe(
  fp.when(fp.isObject, fp.pipe(fp.keys, fp.first))
)

export const getCommandFromMessage = (message) =>
  fp.find(
    fp.pipe(
      fp.get('text'),
      fp.isEqual(fp.pipe(fp.split(' '), fp.first)(message))
    )
  )

export const getCommand = (message, language, tenant) => {
  const feature = getFeature(tenant)
  const commandGroup = fp.omitBy((value) => {
    const flowName = fp.pipe(fp.split(':'), fp.first)(value)
    const applicationName = fp.pipe(
      fp.split(':'),
      fp.first
    )(feature.application)
    const allowGroup = [applicationName, 'common']

    return !allowGroup.includes(flowName)
  }, commands)

  const command = fp.pipe(
    fp.values,
    fp.map((value) => ({ key: value, text: i18n.getFixedT(language)(value) })),
    getCommandFromMessage(message),
    fp.get('key')
  )(commandGroup)

  if (command) {
    const args = fp.pipe(fp.split(' '), fp.slice(1, Infinity))(message)

    return [command, ...args]
  }

  return null
}

export const getMeta = (state) => fp.values(state.meta)[0]

export const getAnswer = fp.pipe(
  fp.props(['content.text']),
  fp.find(fp.complement(fp.isNil))
)

export const updateData = (key, fn) => (context, event) => ({
  data: fp.set(
    key.replace('data.', ''),
    fp.isFunction(fn) ? fn(event.payload) : getAnswer(event.payload),
    context.data
  ),
})

export const updateSession = (key, fn) => (context, event) =>
  fp.set(key, fn(event), {})

export const getMessageTemplate = ({
  action = {},
  event = {},
  session = {},
  dependencies = {},
} = {}) =>
  fp.pipe(
    fp.concat([]),
    fp.map(
      fp.when(fp.isFunction, (fn) =>
        fn({
          action,
          event,
          session,
          dependencies,
        }))
    ),
    fp.flatten,
    fp.reject(fp.isNil),
    fp.map(
      fp.traverse(
        fp.when(
          fp.pipe(fp.nthArg(1), fp.last, fp.includes(fp.__, translationKeys)),
          (value) => {
            return i18n.getFixedT(session.language, session.tenant)(
              value,
              {
                action,
                event,
                session,
                dependencies,
                interpolation: { escapeValue: false },
              }
            )
          }
        )
      )
    ),
    fp.reject(fp.anyPass(fp.isNil, fp.isEmpty))
  )

export const clearCurrentForm = async (caseId, tenant) => {
  try {
    const input = {
      id: caseId,
      currentForm: null,
      /** NOTE: this key will be used to validate each LIFF whether the client can open the form or not */
      latestFlowStartedAt: new Date().toISOString(),
    }
    await appSyncClient.mutate({
      mutation: updateCaseMutation,
      variables: { input, tenant },
    })
    // eslint-disable-next-line no-console
    logger.info(`cleared currentForm of case:${caseId}`)
    return
  } catch (error) {
    // eslint-disable-next-line no-console
    logger.info('Error update case', error)
  }
}
