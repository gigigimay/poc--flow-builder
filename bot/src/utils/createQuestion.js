import * as fp from '@appman/mac_modules/fp'
import { actions } from 'xstate'
import { ANSWER, RESUME, RESUMED } from 'constants/state'
import { getAnswer } from './state'

// set default on.RESUME to go to question
export const addDefaultOnResume = fp.update(
  'on',
  fp.flip(fp.defaultsDeep)({
    [RESUME]: {
      actions: actions.assign({ status: RESUMED }),
      target: '.question',
    },
  })
)

export const mapStateWithQuestion = (state) => {
  const hasQuestion = fp.has('states.question')(state)
  if (!hasQuestion) {
    return state
  }
  return fp.pipe(
    // set initial state to `question`
    fp.set('initial', 'question'),
    addDefaultOnResume
  )(state)
}

export const toArray = fp.concat([])

// if `cond` is an array, map it into multiple answer conditions
export const flattenConditionArray = fp.pipe(
  fp.map(
    fp.when(
      fp.pipe(fp.get('cond'), fp.isArray),
      ({ cond, ...action }) =>
        fp.map((cond) => ({ cond, ...action }), cond)
    )
  ),
  fp.flatten,
)

// map text from ANSWER event into `cond` key of on.ANSWER
export const mapTextToCondition = fp.map(
  fp.update(
    'cond',
    fp.when(
      fp.isFunction,
      (fn) => (context, event) =>
        fn(getAnswer(event.payload), context, event)
    )
  )
)

export const mapStateAnswer = (state) => {
  const onAnswer = fp.get(['on', ANSWER])(state)
  if (fp.isNil(onAnswer)) {
    return state
  }
  const newAnswer = fp.pipe(
    toArray,
    flattenConditionArray,
    mapTextToCondition,
  )(onAnswer)
  return fp.set(['on', ANSWER], newAnswer)(state)
}

const mapStates = (states) => fp.mapValues(
  fp.pipe(
    mapStateWithQuestion,
    mapStateAnswer,
    fp.when(fp.has('states'), fp.update('states', mapStates))
  )
)(states)

const createQuestion = (config) => {
  return fp.pipe(
    fp.set('entry', actions.assign({ status: RESUMED })),
    fp.update('states', mapStates)
  )(config)
}

export default createQuestion
