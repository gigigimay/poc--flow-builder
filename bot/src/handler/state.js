import * as fp from '@appman/mac_modules/fp'
import { Machine, actions } from 'xstate'
import {
  isTextMessage,
  getModule,
  getMeta,
  getMessageTemplate,
  clearCurrentForm,
} from 'utils/state'
import {
  INVALID,
  NEXT,
  COMMAND,
  RESUME,
  PAUSED,
  ERROR,
  BROKEN,
  FALLBACK,
} from 'constants/state'

import stateConfig from 'config/states'
import { getLogger } from 'utils/logger'
import { getFeature } from 'config/features'

const logger = getLogger('handler/state.js')

const guards = {
  isTextMessage: fp.pipe(fp.nthArg(1), fp.get('payload'), isTextMessage),
  isNotTextMessage: fp.pipe(
    fp.nthArg(1),
    fp.get('payload'),
    fp.complement(isTextMessage)
  ),
}

const getStateConfig = (stateConfig) =>
  fp.assign(stateConfig.states, {
    listening: {
      entry: actions.assign({ status: PAUSED, finished: true }),
      on: stateConfig.events,
    },
  })

export const listen = async (
  state,
  { action = {}, event, session, dependencies }
) => {
  if (!fp.isEmpty(state.meta) && !fp.isEmpty(event)) {
    const meta = getMeta(state)

    if (meta.message) {
      const message = getMessageTemplate({
        action,
        event,
        session,
        dependencies,
      })(meta.message)

      if (!fp.isEmpty(message)) {
        return message
      }
    }
  }

  return null
}

export const handleDependencies = async ({
  dependencies,
  actions,
  event,
  session,
}) => {
  const handleDependency = async (dependency) => {
    if (fp.isFunction(dependency)) {
      return dependency({
        event,
        session,
      })
    }
    if (fp.isString(dependency) && fp.isFunction(actions[dependency])) {
      const fn = actions[dependency]
      return fn({ session, event })
    }
    return undefined
  }

  if (Array.isArray(dependencies)) {
    const dependencyResults = await Promise.all(
      dependencies.map(handleDependency)
    )
    return dependencyResults.reduce(Object.assign, {})
  }
  return handleDependency(dependencies)
}

export default async (
  type,
  action = {},
  { session, event, actions = {} } = {},
  tenant
) => {
  const {
    // application,
    module: prevModule,
    intent: prevIntent,
    status: prevStatus,
  } = session

  const feature = getFeature(tenant)
  const states = getStateConfig(stateConfig[feature.application])

  let stateMachine = Machine(
    {
      id: 'state',
      initial: 'listening',
      context: {
        finished: false,
        data: {},
      },
      states,
    },
    {
      guards,
      actions,
    }
  )

  stateMachine = stateMachine.withContext(
    fp.defaults(stateMachine.context, session)
  )

  let currentState = fp.getOr(
    stateMachine.initialState.value,
    prevModule,
    prevIntent
  )

  const processMeta = async () => {
    try {
      await Promise.all(
        currentState.actions.map(
          (action) => action.exec && action.exec(currentState.context, event)
        )
      )

      const dependencies = fp.get('dependencies', getMeta(currentState))

      action.dependencies = await handleDependencies({
        dependencies,
        actions,
        event,
        session: currentState.context,
      })

      const transition = fp.get('transition', getMeta(currentState))
      const isExit = fp.isEqual(
        fp.last(currentState.toStrings()),
        fp.get(prevModule, prevIntent)
      )
      if (fp.isFunction(transition)) {
        const nextEvent = await transition({
          isExit,
          event,
          dependencies: action.dependencies,
          session: currentState.context,
        })
        if (nextEvent) {
          currentState = stateMachine.transition(currentState, nextEvent)
          await processMeta()
        }
      }
    } catch (e) {
      currentState = stateMachine.transition(currentState, INVALID)
      logger.error(`Error processing meta: ${e}\n${JSON.stringify(e, null, 2)}`)
    }
  }

  const skipQuestion = async () => {
    let field = null

    await processMeta()

    do {
      const intent = fp.last(currentState.toStrings())

      if (fp.endsWith('question', intent)) {
        field = fp.get('field', getMeta(currentState))
      } else {
        field = null
      }

      if (field) {
        const fieldData = fp.get(field, currentState.context)

        if (fp.isNil(fieldData)) {
          field = null
        } else {
          currentState = stateMachine.transition(currentState, NEXT)

          // eslint-disable-next-line no-await-in-loop
          await processMeta()

          field = fp.get('field', getMeta(currentState))
        }
      }
    } while (!fp.isNil(field) && fp.contains(NEXT, currentState.nextEvents))
  }

  if (type === COMMAND) {
    // NOTE: start the flow / switch across each flow

    if (session.caseId > 0) {
      await clearCurrentForm(session.caseId, tenant)
      // clear values of field 'data' in context
      stateMachine.context.data = {}
    }

    const initialState = stateMachine.transition(
      stateMachine.initialState.value,
      {
        type: action.payload,
        ...action,
      }
    )
    currentState = initialState

    await skipQuestion()
  } else if (type === FALLBACK && states[FALLBACK]) {
    /** NOTE: enter fallback flow when the client starts conversation with other messages than a command */

    logger.info(`Redirecting chatId:${session.id} to FALLBACK flow`)
    const initialState = stateMachine.transition(
      stateMachine.initialState.value,
      {
        type: FALLBACK,
        ...action,
      }
    )
    currentState = initialState

    await skipQuestion()
  } else if (type === ERROR) {
    const initialState = stateMachine.transition(
      stateMachine.initialState.value,
      {
        type,
        ...action,
      }
    )
    await listen(initialState, {
      action,
      event,
    })

    return { messages: null, session }
  } else if (prevStatus === BROKEN) {
    currentState = stateMachine.transition(currentState, RESUME)

    await skipQuestion()
  } else if (prevStatus === PAUSED) {
    return { messages: null, session }
  } else {
    currentState = stateMachine.transition(currentState, { type, ...action })
    await skipQuestion()
  }

  const { status, finished = false, ...updatedSession } = currentState.context

  const nextState = fp.last(currentState.toStrings())

  let currentStatus = status
  let currentModule = getModule(currentState.value)
  let currentIntent = nextState

  let messages
  try {
    messages = await listen(currentState, {
      action,
      event,
      dependencies: action.dependencies,
      session: currentState.context,
    })
  } catch (e) {
    currentStatus = BROKEN
    logger.error(`Error getting messages: ${e}\n${JSON.stringify(e, null, 2)}`)
  }

  if (finished) {
    currentIntent = fp.omit(currentModule, prevIntent)
    currentModule = ''
    // TODO - fix later since it is not safe
    updatedSession.data = {}
  } else {
    currentIntent = fp.when(
      () => !fp.isEmpty(currentModule),
      fp.set(currentModule, currentIntent)
    )(prevIntent)
  }

  return {
    messages,
    session: {
      ...updatedSession,
      status: currentStatus,
      intent: currentIntent,
      module:
        currentModule === stateMachine.initialState.value ? '' : currentModule,
    },
  }
}
