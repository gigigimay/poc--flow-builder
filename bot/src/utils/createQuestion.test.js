import * as fp from '@appman/mac_modules/fp'
import { text } from 'config/templates'
import { isAnswer } from 'config/validation'
import { ANSWER, NEXT, RESUME, RESUMED, SKIP } from 'constants/state'
import { actions } from 'xstate'
import { updateData } from './state'
import createQuestion, {
  addDefaultOnResume,
  toArray,
  flattenConditionArray,
  mapTextToCondition,
  mapStateWithQuestion,
  mapStateAnswer,
} from './createQuestion'

describe('addDefaultOnResume', () => {
  it('should add default on.RESUME to state', () => {
    const state = {}
    expect(addDefaultOnResume(state)).toEqual({
      on: {
        [RESUME]: {
          actions: actions.assign({ status: RESUMED }),
          target: '.question',
        },
      },
    })
  })
  it('should prefer the state over the default value', () => {
    const state = {
      on: {
        [ANSWER]: 'something',
        [RESUME]: {
          target: '.question2',
        },
      },
    }
    expect(addDefaultOnResume(state)).toEqual({
      on: {
        [ANSWER]: 'something',
        [RESUME]: {
          actions: actions.assign({ status: RESUMED }),
          target: '.question2',
        },
      },
    })
  })
})

describe('mapStateWithQuestion', () => {
  it('should return state if it does not have question', () => {
    const state = {}
    expect(mapStateWithQuestion(state)).toEqual({})
  })
  it('should return updated state if it has question', () => {
    const state = {
      states: {
        question: {},
      },
    }
    expect(mapStateWithQuestion(state)).toEqual({
      initial: 'question',
      states: {
        question: {},
      },
      on: {
        [RESUME]: {
          actions: actions.assign({ status: RESUMED }),
          target: '.question',
        },
      },
    })
  })
})

describe('toArray', () => {
  it('should return [] when input is []', () => {
    expect(toArray([])).toEqual([])
  })
  it('should return [{}] when input is {}', () => {
    expect(toArray({})).toEqual([{}])
  })
})

describe('flattenConditionArray', () => {
  it('should map array of `cond` to multiple conditions', () => {
    const answers = [
      { cond: [1, 2, 3], target: 'A' },
      { cond: 4, target: 'B' },
    ]
    expect(flattenConditionArray(answers)).toEqual([
      { cond: 1, target: 'A' },
      { cond: 2, target: 'A' },
      { cond: 3, target: 'A' },
      { cond: 4, target: 'B' },
    ])
  })
})

describe('mapTextToCondFunction', () => {
  it('should', () => {
    const cond = (text, context, event) => ({ text, context, event })
    const answers = [{ cond }]
    const result = mapTextToCondition(answers)
    const context = {}
    const event = { payload: { content: { text: 'ABC' } } }
    expect(result[0].cond(context, event)).toEqual(cond('ABC', context, event))
  })
})

describe('mapStateAnswer', () => {
  it('should return state if on.ANSWER is nil', () => {
    const state = {}
    expect(mapStateAnswer(state)).toEqual({})
  })
  it('should return updated state if on.ANSWER is not nil', () => {
    const state = {
      on: {
        [ANSWER]: {},
      },
    }
    expect(mapStateAnswer(state)).toEqual({
      on: {
        [ANSWER]: [{}],
      },
    })
  })
})

describe('createQuestion', () => {
  it('should match snapshots when called with empty object', () => {
    const result = createQuestion({})
    expect(result).toMatchSnapshot(result)
  })

  it('should match snapshots when called with state', () => {
    const mockState = {
      initial: 'show_new_product',
      states: {
        show_new_product: {
          states: {
            question: {
              meta: {
                field: 'data.buy_new_product',
                dependencies: jest.fn(),
                message: jest.fn(),
              },
            },
            invalid: {
              meta: {
                message: text({ text: 'error' }),
              },
            },
          },
          on: {
            [ANSWER]: [
              {
                cond: isAnswer('ซื้ออันนี้'),
                actions: [
                  actions.assign(
                    updateData('data.buy_new_product', fp.constant(12344))
                  ),
                  actions.raise(SKIP),
                ],
              },
              {
                cond: isAnswer('แนะนำอันใหม่'),
                actions: [actions.raise(NEXT)],
              },
            ],
            [NEXT]: {
              target: 'suggest_product',
            },
            [SKIP]: {
              target: 'addition',
            },
          },
        },
        suggest_product: {
          states: {
            question: {
              meta: {
                message: text({ text: '[new] suggest product~~' }),
              },
            },
          },
        },
        addition: {
          states: {
            question: {
              meta: {
                message: text({ text: '[new] addtional question~~' }),
              },
            },
          },
        },
      },
    }

    const result = createQuestion(mockState)
    expect(result).toMatchSnapshot(result)
  })
})
