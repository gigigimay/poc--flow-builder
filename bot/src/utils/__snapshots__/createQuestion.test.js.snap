// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createQuestion should match snapshots when called with empty object 1`] = `
Object {
  "entry": Object {
    "assignment": Object {
      "status": "RESUMED",
    },
    "type": "xstate.assign",
  },
  "states": Object {},
}
`;

exports[`createQuestion should match snapshots when called with state 1`] = `
Object {
  "entry": Object {
    "assignment": Object {
      "status": "RESUMED",
    },
    "type": "xstate.assign",
  },
  "initial": "show_new_product",
  "states": Object {
    "addition": Object {
      "initial": "question",
      "on": Object {
        "RESUME": Object {
          "actions": Object {
            "assignment": Object {
              "status": "RESUMED",
            },
            "type": "xstate.assign",
          },
          "target": ".question",
        },
      },
      "states": Object {
        "question": Object {
          "meta": Object {
            "message": [Function],
          },
        },
      },
    },
    "show_new_product": Object {
      "initial": "question",
      "on": Object {
        "ANSWER": Array [
          Object {
            "actions": Array [
              Object {
                "assignment": [Function],
                "type": "xstate.assign",
              },
              Object {
                "event": "SKIP",
                "type": "xstate.raise",
              },
            ],
            "cond": [Function],
          },
          Object {
            "actions": Array [
              Object {
                "event": "NEXT",
                "type": "xstate.raise",
              },
            ],
            "cond": [Function],
          },
        ],
        "NEXT": Object {
          "target": "suggest_product",
        },
        "RESUME": Object {
          "actions": Object {
            "assignment": Object {
              "status": "RESUMED",
            },
            "type": "xstate.assign",
          },
          "target": ".question",
        },
        "SKIP": Object {
          "target": "addition",
        },
      },
      "states": Object {
        "invalid": Object {
          "meta": Object {
            "message": [Function],
          },
        },
        "question": Object {
          "meta": Object {
            "dependencies": [MockFunction],
            "field": "data.buy_new_product",
            "message": [MockFunction],
          },
        },
      },
    },
    "suggest_product": Object {
      "initial": "question",
      "on": Object {
        "RESUME": Object {
          "actions": Object {
            "assignment": Object {
              "status": "RESUMED",
            },
            "type": "xstate.assign",
          },
          "target": ".question",
        },
      },
      "states": Object {
        "question": Object {
          "meta": Object {
            "message": [Function],
          },
        },
      },
    },
  },
}
`;
